### PR TITLE
DRILL-4047: Select with options

### PR DESCRIPTION
**Pull request by julienledem**
Implementation of passing options down to a table while doing a select.

This depends on the following PR in Drill's Calcite fork:
https://github.com/dremio/calcite/pull/1

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSystemSchemaFactory.java [hunk 1]
```diff
@@ -86,7 +86,8 @@ public FileSystemSchema(String name, SchemaConfig schemaConfig) throws IOExcepti
 
     void setPlus(SchemaPlus plusOfThis){
       for(WorkspaceSchema s : schemaMap.values()){
-        plusOfThis.add(s.getName(), s);
+        SchemaPlus schemaPlus = plusOfThis.add(s.getName(), s);
+//        schemaPlus.add(arg0, arg1);
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FileSystemSchemaFactory.java]: todo: cleanup

File: exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java [hunk 2]
```diff
@@ -86,7 +87,19 @@ public QueryContext(final UserSession session, final DrillbitContext drillbitCon
...
+//            getRootSchema(),
+//            caseSensitive,
+//            getNewDefaultSchema(),
+//            getTypeFactory());
     table = new DrillOperatorTable(getFunctionRegistry());
+//    table = new ChainedSqlOperatorTable(asList(
+//          SqlStdOperatorTable.instance(),
+//          catalogReader,
+//          new DrillOperatorTable(getFunctionRegistry())
+//        ));
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java]: todo: cleanup

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FormatCreator.java [hunk 3]
```diff
@@ -92,10 +118,39 @@
...
+   */
+  FormatPlugin newFormatPlugin(FormatPluginConfig fpconfig) {
+    Constructor<?> c = configConstructors.get(fpconfig.getClass());
+    if (c == null) {
+      throw new RuntimeException("Unable to find constructor for storage config of type " + fpconfig.getClass().getName());
+    }
+    try {
+      return (FormatPlugin) c.newInstance(null, context, fsConf, storageConfig, fpconfig);
+    } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException e1) {
+      throw new RuntimeException("Failure initializing storage config of type " + fpconfig.getClass().getName(), e1);
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/FormatCreator.java]: todo: exception

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 4]
```diff
@@ -221,9 +603,46 @@ public void dropView(String viewName) throws IOException {
...
+    public List<Function> getFunctions(String name) {
+      System.out.println("getFunctions(" + name + ")");
+      List<TableSignature> sigs = optionExtractor.getTableSignatures(name);
+      System.out.println(sigs);
+//      List<TableSignature> sigs = Arrays.asList(
+//          new TableSignature(name, new TableParamDef("delimiter", String.class, true)),
+//          new TableSignature(name, new TableParamDef("delimiter", Integer.TYPE, true)),
+//          new TableSignature(name, new TableParamDef("foo", Integer.TYPE, true), new TableParamDef("bar", Integer.TYPE, true)),
+//          new TableSignature(name, new TableParamDef("foo", String.class, true))
+//      );
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: todo:cleanup

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 5]
```diff
@@ -221,9 +603,46 @@ public void dropView(String viewName) throws IOException {
...
+
+    @Override
+    public Set<String> getFunctionNames() {
+      System.out.println("getFunctionNames");
+      return rawTableNames();
+    }
+
+    @Override
+    public List<Function> getFunctions(String name) {
+      System.out.println("getFunctions(" + name + ")");
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: todo: cleanup

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 6]
```diff
@@ -221,9 +603,46 @@ public void dropView(String viewName) throws IOException {
...
     @Override
     public Set<String> getTableNames() {
-      return Sets.union(tables.keySet(), getViews());
+      System.out.println("getTableNames");
+      return Sets.union(rawTableNames(), getViews());
+    }
+
+    @Override
+    public Set<String> getFunctionNames() {
+      System.out.println("getFunctionNames");
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: todo: cleanup

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 7]
```diff
@@ -221,9 +603,46 @@ public void dropView(String viewName) throws IOException {
...
+        public String apply(TableInstance input) {
+          return input.sig.name;
+        }
+      }));
+    }
+
     @Override
     public Set<String> getTableNames() {
-      return Sets.union(tables.keySet(), getViews());
+      System.out.println("getTableNames");
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: todo: cleanup

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 8]
```diff
@@ -313,19 +734,29 @@ public String getTypeName() {
...
+        FileSelection fileSelection = FileSelection.create(fs, config.getLocation(), key.sig.name);
         if (fileSelection == null) {
           return null;
         }
-
+        if (key.sig.params.size() > 0) {
+          FormatPluginConfig fconfig = optionExtractor.eval(key);
+//          TextFormatPlugin.TextFormatConfig fconfig = new TextFormatPlugin.TextFormatConfig();
+//          fconfig.extensions = Arrays.asList();
+//          fconfig.fieldDelimiter = ((String)key.params.get(0)).charAt(0);
```
Comment (julienledem) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: todo: cleanup

File: exec/java-exec/src/test/java/org/apache/drill/TestSelectWithOption.java [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+      }
+    }
+
+    String query = "select columns from table(dfs.`${WORKING_PATH}/" + input.getPath() +
+        "`(type => 'PARQUET'))";
+    System.out.println(query);
+
+    test(query);
+
+  }
```
Comment (julienledem) [inline exec/java-exec/src/test/java/org/apache/drill/TestSelectWithOption.java]: todo more granular tests

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 10]
```diff
@@ -148,12 +175,427 @@ private Path getViewPath(String name) {
...
     return new WorkspaceSchema(parentSchemaPath, schemaName, schemaConfig);
   }
 
-  public class WorkspaceSchema extends AbstractSchema implements ExpandingConcurrentMap.MapValueFactory<String, DrillTable> {
-    private final ExpandingConcurrentMap<String, DrillTable> tables = new ExpandingConcurrentMap<>(this);
+  /**
+   * Describes the options for a format plugin
+   * extracted from the FormatPluginConfig subclass
+   */
+  static final class OptionsDescriptor {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Can you pull this out into a separate file?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 11]
```diff
@@ -148,12 +175,427 @@ private Path getViewPath(String name) {
...
+              .addContext("table", t.sig.name)
+              .addContext("parameter", paramDef.name)
+              .build(logger);
+        }
+      }
+      return config;
+    }
+
+    private void badInput(TableInstance t) {
+      throw new IllegalArgumentException("The parameters provided are not applicable to the type specified:\n"
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: UserException?

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 12]
```diff
@@ -148,12 +175,427 @@ private Path getViewPath(String name) {
...
+            .build(logger);
+      }
+      for (int i = 1; i < t.params.size(); i++) {
+        Object param = t.params.get(i);
+        if (param == null)
+          continue;
+        TableParamDef paramDef = t.sig.params.get(i);
+        TableParamDef expectedParamDef = this.functionParamsByName.get(paramDef.name);
+        if (expectedParamDef == null || expectedParamDef.type != paramDef.type) {
+          badInput(t);
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: I think you should make these errors much more specific (e.g. param 'name' was of type 'blah' but was actually of type 'blah'. Expected parameters are: ...

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java [hunk 13]
```diff
@@ -148,12 +175,427 @@ private Path getViewPath(String name) {
...
+        if (i != 0) {
+          sb.append(", ");
+        }
+        sb.append(paramDef.name).append(": ").append(paramDef.type.getSimpleName());
+      }
+      sb.append(")");
+      return sb.toString();
+    }
+
+    FormatPluginConfig eval(TableInstance t) {
```
Comment (jacques-n) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/WorkspaceSchemaFactory.java]: Can you provide a more descriptive method name and short doc?

File: exec/java-exec/src/test/java/org/apache/drill/TestSelectWithOption.java [hunk 14]
```diff
@@ -0,0 +1,203 @@
...
+  @Test
+  public void testVariationsCSV() throws Exception {
+    String csvTableName = genCSVTable("testVariationsCSV",
+        "a,b",
+        "c|d");
+    // Using the defaults in TextFormatConfig (the field delimiter is neither "," not "|")
+    String[] csvQueries = {
+//        format("select columns from %s ('TeXT')", csvTableName),
+//        format("select columns from %s('TeXT')", csvTableName),
+        format("select columns from table(%s ('TeXT'))", csvTableName),
```
Comment (jacques-n) [inline exec/java-exec/src/test/java/org/apache/drill/TestSelectWithOption.java]: Do you have a test where data it isn't TeXT?

Comment (julienledem) [inline exec/java-exec/src/test/java/org/apache/drill/TestSelectWithOption.java]: no, but the point was to use a capitalization that is not used anywhere in the code to make sure this was case insensitive. I can vary the spelling but I thought this was sufficient.

**Comment by julienledem** [comment]
I ran the full test suite. It's green

**Pull request closed without merging**